File: _may_messenger_web_client/src/services/offlineSyncService.ts
import { outboxRepository } from '../repositories/outboxRepository';
import { messageApi } from '../api/messageApi';
import { Message, MessageStatus, MessageType } from '../types/chat';
import { PendingMessage } from './indexedDBStorage';

/**
 * Background service for syncing pending messages from outbox
 */
class OfflineSyncService {
  private syncTimer: NodeJS.Timeout | null = null;
  private isSyncing = false;
  
  onMessageSynced?: (localId: string, serverId: string) => void;
  onMessageStatusUpdate?: (localId: string, status: MessageStatus) => void;

  /**
   * Start the background sync service
   */
  start(): void {
    console.log('[SYNC] Starting offline sync service');
    
    // Listen for online/offline events
    window.addEventListener('online', this.handleOnline);
    window.addEventListener('offline', this.handleOffline);
    
    // Periodic sync every 30 seconds
    this.syncTimer = setInterval(() => {
      this.syncNow();
    }, 30000);
    
    // Initial sync
    this.syncNow();
  }

  private handleOnline = (): void => {
    console.log('[SYNC] Network connected, triggering sync');
    this.syncNow();
  };

  private handleOffline = (): void => {
    console.log('[SYNC] Network disconnected');
  };

  /**
   * Manually trigger sync
   */
  async syncNow(): Promise<void> {
    if (this.isSyncing) {
      console.log('[SYNC] Already syncing, skipping');
      return;
    }
    
    this.isSyncing = true;
    
    try {
      // Check if online
      if (!navigator.onLine) {
        console.log('[SYNC] No network connection, skipping sync');
        return;
      }
      
      // Get all pending messages
      const pendingMessages = await outboxRepository.getAllPendingMessages();
      
      if (pendingMessages.length === 0) {
        console.log('[SYNC] No pending messages to sync');
        return;
      }
      
      console.log(`[SYNC] Syncing ${pendingMessages.length} pending messages`);
      
      for (const pending of pendingMessages) {
        // Skip messages that are currently syncing
        if (pending.syncState === 'syncing') {
          continue;
        }
        
        // Check if message should be retried (with exponential backoff)
        if (pending.syncState === 'failed') {
          const shouldRetry = this.shouldRetry(pending.retryCount, new Date(pending.createdAt));
          if (!shouldRetry) {
            console.log(`[SYNC] Skipping message ${pending.localId} (retry limit or cooldown)`);
            continue;
          }
        }
        
        // Attempt to sync message
        await this.syncMessage(pending);
      }
      
    } catch (error) {
      console.error('[SYNC] Sync error:', error);
    } finally {
      this.isSyncing = false;
    }
  }

  /**
   * Sync a single pending message
   */
  private async syncMessage(pending: PendingMessage): Promise<void> {
    try {
      console.log(`[SYNC] Syncing message: ${pending.localId} (attempt ${pending.retryCount + 1})`);
      
      // Mark as syncing
      await outboxRepository.markAsSyncing(pending.localId);
      
      // Send to backend
      let serverMessage: Message;
      if (pending.type === MessageType.Text) {
        serverMessage = await messageApi.sendMessage({
          chatId: pending.chatId,
          type: pending.type,
          content: pending.content,
        });
      } else {
        // Audio messages need special handling
        // For now, mark as failed if we don't have the blob
        throw new Error('Audio message sync not implemented');
      }
      
      console.log(`[SYNC] Message synced successfully: ${pending.localId} -> ${serverMessage.id}`);
      
      // Mark as synced
      await outboxRepository.markAsSynced(pending.localId, serverMessage.id);
      
      // Notify listeners
      this.onMessageSynced?.(pending.localId, serverMessage.id);
      this.onMessageStatusUpdate?.(pending.localId, MessageStatus.Sent);
      
      // Clean up after successful sync (with delay)
      setTimeout(() => {
        outboxRepository.removePendingMessage(pending.localId);
      }, 120000); // 2 minutes
      
    } catch (error: any) {
      console.error(`[SYNC] Failed to sync message ${pending.localId}:`, error);
      
      // Mark as failed
      const errorMessage = error.response?.data?.message || error.message || 'Sync failed';
      await outboxRepository.markAsFailed(pending.localId, errorMessage);
      
      // Notify listeners
      this.onMessageStatusUpdate?.(pending.localId, MessageStatus.Failed);
    }
  }

  /**
   * Determine if a failed message should be retried (exponential backoff)
   */
  private shouldRetry(retryCount: number, createdAt: Date): boolean {
    const maxRetries = 10;
    
    if (retryCount >= maxRetries) {
      return false;
    }
    
    // Exponential backoff: 10s, 20s, 40s, 80s, 160s, ...
    const backoffSeconds = 10 * Math.pow(2, retryCount);
    const nextRetryTime = new Date(createdAt.getTime() + backoffSeconds * 1000);
    
    return new Date() > nextRetryTime;
  }

  /**
   * Retry a specific failed message immediately
   */
  async retryMessage(localId: string): Promise<void> {
    console.log('[SYNC] Manual retry requested for message:', localId);
    
    const message = await outboxRepository.getPendingMessageById(localId);
    if (!message) {
      console.log('[SYNC] Message not found:', localId);
      return;
    }
    
    if (message.syncState !== 'failed') {
      console.log('[SYNC] Message is not in failed state:', localId);
      return;
    }
    
    // Reset to local-only state
    await outboxRepository.retryMessage(localId);
    
    // Trigger sync
    await this.syncMessage(message);
  }

  /**
   * Get statistics about pending messages
   */
  async getStats(): Promise<{
    total: number;
    pending: number;
    syncing: number;
    failed: number;
    synced: number;
  }> {
    const all = await outboxRepository.getAllPendingMessages();
    return {
      total: all.length,
      pending: all.filter((m) => m.syncState === 'localOnly').length,
      syncing: all.filter((m) => m.syncState === 'syncing').length,
      failed: all.filter((m) => m.syncState === 'failed').length,
      synced: all.filter((m) => m.syncState === 'synced').length,
    };
  }

  /**
   * Stop the sync service
   */
  stop(): void {
    console.log('[SYNC] Stopping offline sync service');
    
    if (this.syncTimer) {
      clearInterval(this.syncTimer);
      this.syncTimer = null;
    }
    
    window.removeEventListener('online', this.handleOnline);
    window.removeEventListener('offline', this.handleOffline);
  }

  dispose(): void {
    this.stop();
  }
}

export const offlineSyncService = new OfflineSyncService();

File: _may_messenger_mobile_app/lib/data/services/offline_sync_service.dart
import 'dart:async';
import 'package:connectivity_plus/connectivity_plus.dart';
import '../repositories/outbox_repository.dart';
import '../repositories/message_repository.dart';
import '../models/message_model.dart';

/// Background service for syncing pending messages from outbox
class OfflineSyncService {
  final OutboxRepository _outboxRepository;
  final MessageRepository _messageRepository;
  final Connectivity _connectivity = Connectivity();
  
  Timer? _syncTimer;
  StreamSubscription? _connectivitySubscription;
  bool _isSyncing = false;
  
  Function(String localId, String serverId)? onMessageSynced;
  Function(String localId, MessageStatus status)? onMessageStatusUpdate;

  OfflineSyncService(this._outboxRepository, this._messageRepository);

  /// Start the background sync service
  Future<void> start() async {
    print('[SYNC] Starting offline sync service');
    
    // Listen for connectivity changes
    _connectivitySubscription = _connectivity.onConnectivityChanged.listen((result) {
      if (result != ConnectivityResult.none) {
        print('[SYNC] Network connected, triggering sync');
        syncNow();
      } else {
        print('[SYNC] Network disconnected');
      }
    });
    
    // Periodic sync every 30 seconds
    _syncTimer = Timer.periodic(const Duration(seconds: 30), (timer) {
      syncNow();
    });
    
    // Initial sync
    syncNow();
  }

  /// Manually trigger sync
  Future<void> syncNow() async {
    if (_isSyncing) {
      print('[SYNC] Already syncing, skipping');
      return;
    }
    
    _isSyncing = true;
    
    try {
      // Check if online
      final connectivityResult = await _connectivity.checkConnectivity();
      if (connectivityResult == ConnectivityResult.none) {
        print('[SYNC] No network connection, skipping sync');
        return;
      }
      
      // Get all pending messages
      final pendingMessages = await _outboxRepository.getAllPendingMessages();
      
      if (pendingMessages.isEmpty) {
        print('[SYNC] No pending messages to sync');
        return;
      }
      
      print('[SYNC] Syncing ${pendingMessages.length} pending messages');
      
      for (final pending in pendingMessages) {
        // Skip messages that are currently syncing
        if (pending.syncState == SyncState.syncing) {
          continue;
        }
        
        // Check if message should be retried (with exponential backoff)
        if (pending.syncState == SyncState.failed) {
          final shouldRetry = _shouldRetry(pending.retryCount, pending.createdAt);
          if (!shouldRetry) {
            print('[SYNC] Skipping message ${pending.localId} (retry limit or cooldown)');
            continue;
          }
        }
        
        // Attempt to sync message
        await _syncMessage(pending);
      }
      
    } catch (e) {
      print('[SYNC] Sync error: $e');
    } finally {
      _isSyncing = false;
    }
  }

  /// Sync a single pending message
  /// Supports all message types: text, audio, image, file
  Future<void> _syncMessage(PendingMessage pending) async {
    try {
      print('[SYNC] Syncing ${pending.type.name} message: ${pending.localId} (attempt ${pending.retryCount + 1})');
      
      // Mark as syncing
      await _outboxRepository.markAsSyncing(pending.localId);
      
      // Send to backend based on message type
      Message serverMessage;
      
      switch (pending.type) {
        case MessageType.text:
          serverMessage = await _messageRepository.sendMessage(
            chatId: pending.chatId,
            type: pending.type,
            content: pending.content,
            replyToMessageId: pending.replyToMessageId,
          );
          break;
          
        case MessageType.audio:
          if (pending.localAudioPath == null) {
            throw Exception('Audio path is null for audio message');
          }
          serverMessage = await _messageRepository.sendAudioMessage(
            chatId: pending.chatId,
            audioPath: pending.localAudioPath!,
            replyToMessageId: pending.replyToMessageId,
          );
          break;
          
        case MessageType.image:
          if (pending.localImagePath == null) {
            throw Exception('Image path is null for image message');
          }
          serverMessage = await _messageRepository.sendImageMessage(
            chatId: pending.chatId,
            imagePath: pending.localImagePath!,
            replyToMessageId: pending.replyToMessageId,
          );
          break;
          
        case MessageType.file:
          if (pending.localFilePath == null) {
            throw Exception('File path is null for file message');
          }
          serverMessage = await _messageRepository.sendFileMessage(
            chatId: pending.chatId,
            filePath: pending.localFilePath!,
            fileName: pending.originalFileName ?? 'file',
            replyToMessageId: pending.replyToMessageId,
          );
          break;
          
        case MessageType.poll:
          throw Exception('Poll messages should not be in offline sync');
          
        case MessageType.video:
          throw Exception('Video messages should not be in offline sync');
      }
      
      print('[SYNC] Message synced successfully: ${pending.localId} -> ${serverMessage.id}');
      
      // Mark as synced
      await _outboxRepository.markAsSynced(pending.localId, serverMessage.id);
      
      // Notify listeners
      onMessageSynced?.call(pending.localId, serverMessage.id);
      onMessageStatusUpdate?.call(pending.localId, MessageStatus.sent);
      
      // Clean up after successful sync (with delay)
      Future.delayed(const Duration(minutes: 2), () {
        _outboxRepository.removePendingMessage(pending.localId);
      });
      
    } catch (e) {
      print('[SYNC] Failed to sync message ${pending.localId}: $e');
      
      // Mark as failed
      await _outboxRepository.markAsFailed(pending.localId, e.toString());
      
      // Notify listeners
      onMessageStatusUpdate?.call(pending.localId, MessageStatus.failed);
    }
  }

  /// Determine if a failed message should be retried (exponential backoff)
  bool _shouldRetry(int retryCount, DateTime createdAt) {
    const maxRetries = 10;
    
    if (retryCount >= maxRetries) {
      return false;
    }
    
    // Exponential backoff: 10s, 20s, 40s, 80s, 160s, ...
    final backoffSeconds = 10 * (1 << retryCount);
    final nextRetryTime = createdAt.add(Duration(seconds: backoffSeconds));
    
    return DateTime.now().isAfter(nextRetryTime);
  }

  /// Retry a specific failed message immediately
  Future<void> retryMessage(String localId) async {
    print('[SYNC] Manual retry requested for message: $localId');
    
    final message = await _outboxRepository.getPendingMessageById(localId);
    if (message == null) {
      print('[SYNC] Message not found: $localId');
      return;
    }
    
    if (message.syncState != SyncState.failed) {
      print('[SYNC] Message is not in failed state: $localId');
      return;
    }
    
    // Reset to local-only state
    await _outboxRepository.retryMessage(localId);
    
    // Trigger sync
    await _syncMessage(message);
  }

  /// Get statistics about pending messages
  Future<Map<String, int>> getStats() async {
    final all = await _outboxRepository.getAllPendingMessages();
    return {
      'total': all.length,
      'pending': all.where((m) => m.syncState == SyncState.localOnly).length,
      'syncing': all.where((m) => m.syncState == SyncState.syncing).length,
      'failed': all.where((m) => m.syncState == SyncState.failed).length,
      'synced': all.where((m) => m.syncState == SyncState.synced).length,
    };
  }

  /// Stop the sync service
  void stop() {
    print('[SYNC] Stopping offline sync service');
    _syncTimer?.cancel();
    _syncTimer = null;
    _connectivitySubscription?.cancel();
    _connectivitySubscription = null;
  }

  void dispose() {
    stop();
  }
}

